pkg/execution/debounce/debounce.go

```diff
@@ -0,0 +1,362 @@
+	return DebouncePayload{
+		AccountID:       d.AccountID,
+		WorkspaceID:     d.WorkspaceID,
+		FunctionID:      d.FunctionID,
+		FunctionVersion: d.FunctionVersion,
+	}
+}
+
+func (d DebounceItem) GetInternalID() ulid.ULID {
+	return d.EventID
+}
+
+func (d DebounceItem) GetEvent() event.Event {
+	return d.Event
+}
+
+// DebouncePayload represents the data stored within the queue's payload.
+type DebouncePayload struct {
+	DebounceID ulid.ULID `json:"debounceID"`
+	// AccountID represents the account for the debounce item
+	AccountID uuid.UUID `json:"aID"`
+	// WorkspaceID represents the account for the debounce item
+	WorkspaceID uuid.UUID `json:"wsID"`
+	// FunctionID represents the function ID that this debounce is for.
+	FunctionID uuid.UUID `json:"fnID"`
+	// FunctionVersion represents the version of the function that was debounced.
+	FunctionVersion int `json:"fnV"`
+}
+
+// Debouncer represents an implementation-agnostic function debouncer, delaying function runs
+// until a specific time period passes when no more events matching a key are received.
+type Debouncer interface {
+	Debounce(ctx context.Context, d DebounceItem, fn inngest.Function) error
+	GetDebounceItem(ctx context.Context, debounceID ulid.ULID) (*DebounceItem, error)
+	DeleteDebounceItem(ctx context.Context, debounceID ulid.ULID) error
+}
+
+func NewRedisDebouncer(r rueidis.Client, k redis_state.DebounceKeyGenerator, q redis_state.QueueManager) Debouncer {
+	return debouncer{
+		r: r,
+		k: k,
+		q: q,
+	}
+}
+
+type debouncer struct {
+	r rueidis.Client
+	k redis_state.DebounceKeyGenerator
+	q redis_state.QueueManager
+}
+
+// DeleteDebounceItem returns a DebounceItem given a debounce ID.
+func (d debouncer) DeleteDebounceItem(ctx context.Context, debounceID ulid.ULID) error {
+	keyDbc := d.k.Debounce(ctx)
+	cmd := d.r.B().Hdel().Key(keyDbc).Field(debounceID.String()).Build()
+	err := d.r.Do(ctx, cmd).Error()
+	if rueidis.IsRedisNil(err) {
+		return nil
+	}
+	return fmt.Errorf("error removing debounce: %w", err)
+}
+
+// GetDebounceItem returns a DebounceItem given a debounce ID.
+func (d debouncer) GetDebounceItem(ctx context.Context, debounceID ulid.ULID) (*DebounceItem, error) {
+	keyDbc := d.k.Debounce(ctx)
+
+	cmd := d.r.B().Hget().Key(keyDbc).Field(debounceID.String()).Build()
+	byt, err := d.r.Do(ctx, cmd).AsBytes()
+	if rueidis.IsRedisNil(err) {
+		return nil, ErrDebounceNotFound
+	}
+
+	di := &DebounceItem{}
+	if err := json.Unmarshal(byt, &di); err != nil {
+		return nil, fmt.Errorf("error unmarshalling debounce item: %w", err)
+	}
+	return di, nil
+}
+
+// Debounce debounces a given function with the given DebounceItem.
+func (d debouncer) Debounce(ctx context.Context, di DebounceItem, fn inngest.Function) error {
+	if fn.Debounce == nil {
+		return fmt.Errorf("fn has no debounce config")
+	}
+	ttl, err := str2duration.ParseDuration(fn.Debounce.Period)
+	if err != nil {
+		return fmt.Errorf("invalid debounce duration: %w", err)
+	}
```

Comment:
Do we want any maximums here? Thinking that a debounce period higher than a particular limit should probably fail at the time of registration.

Reply:
Good idea, yeah, I'll add some consts.